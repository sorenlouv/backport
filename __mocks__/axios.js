const axios = jest.genMockFromModule('axios');
axios.get = jest.fn(() => {
  throw new Error('Forgot to mock axios.get');
});
axios.post = jest.fn(() => {
  throw new Error('Forgot to mock axios.post');
});
module.exports = axios;
