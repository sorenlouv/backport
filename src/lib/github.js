const axios = require('axios');
const querystring = require('querystring');
const get = require('lodash.get');
const constants = require('./constants');

let accessToken;
function getCommitMessage(message) {
  return message.split('\n')[0].trim();
}

function getCommits(owner, repoName, author) {
  const urlArgs = {
    per_page: 20,
    access_token: accessToken
  };

  if (author) {
    urlArgs.author = author;
    urlArgs.per_page = 5;
  }

  return axios(
    `https://api.github.com/repos/${owner}/${repoName}/commits?${querystring.stringify(
      urlArgs
    )}`
  )
    .catch(handleError)
    .then(res =>
      res.data.map(commit => {
        const message = getCommitMessage(commit.commit.message);
        return {
          message,
          sha: commit.sha
        };
      })
    );
}

function getCommit(owner, repoName, sha) {
  return axios(
    `https://api.github.com/repos/${owner}/${repoName}/commits/${sha}?access_token=${accessToken}`
  )
    .catch(handleError)
    .then(res => ({
      message: res.data.commit.message,
      sha: res.data.sha
    }));
}

function createPullRequest(owner, repoName, payload) {
  return axios
    .post(
      `https://api.github.com/repos/${owner}/${repoName}/pulls?access_token=${accessToken}`,
      payload
    )
    .catch(handleError);
}

function addLabels(owner, repoName, pullNumber, labels) {
  return axios
    .post(
      `https://api.github.com/repos/${owner}/${repoName}/issues/${pullNumber}/labels?access_token=${accessToken}`,
      labels
    )
    .catch(handleError);
}

function getPullRequestByCommit(owner, repoName, commitSha) {
  return axios(
    `https://api.github.com/search/issues?q=repo:${owner}/${repoName}+${commitSha}&access_token=${accessToken}`
  )
    .catch(handleError)
    .then(res => get(res.data.items[0], 'number'));
}

function setAccessToken(_accessToken) {
  accessToken = _accessToken;
}

class GithubError extends Error {
  constructor(message) {
    super();
    Error.captureStackTrace(this, GithubError);
    this.code = constants.GITHUB_ERROR;
    this.message = message;
  }
}

function handleError(e) {
  if (get(e.response, 'data')) {
    throw new GithubError(e.response.data);
  }

  throw e;
}

module.exports = {
  setAccessToken,
  addLabels,
  createPullRequest,
  getCommit,
  getCommits,
  getPullRequestByCommit
};
