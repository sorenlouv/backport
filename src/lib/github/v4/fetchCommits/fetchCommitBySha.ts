import { graphql } from '../../../../graphql/generated';
import { ValidConfigOptions } from '../../../../options/options';
import { BackportError } from '../../../BackportError';
import { isMissingConfigFileException } from '../../../remoteConfig';
import {
  Commit,
  parseSourceCommit,
} from '../../../sourceCommit/parseSourceCommit';
import { GithubV4Exception, getGraphQLClient } from '../client/graphqlClient';

export async function fetchCommitBySha(options: {
  accessToken: string;
  branchLabelMapping?: ValidConfigOptions['branchLabelMapping'];
  githubApiBaseUrlV4?: string;
  repoName: string;
  repoOwner: string;
  sha: string;
  sourceBranch: string;
}): Promise<Commit> {
  const {
    accessToken,
    githubApiBaseUrlV4 = 'https://api.github.com/graphql',
    repoName,
    repoOwner,
    sha,
    sourceBranch,
  } = options;

  const query = graphql(`
    query CommitsBySha($repoOwner: String!, $repoName: String!, $sha: String!) {
      repository(owner: $repoOwner, name: $repoName) {
        object(expression: $sha) {
          __typename
          ...SourceCommitWithTargetPullRequestFragment
        }
      }
    }
  `);

  const variables = { repoOwner, repoName, sha };
  const client = getGraphQLClient({ accessToken, githubApiBaseUrlV4 });
  const result = await client.query(query, variables);

  if (result.error && !isMissingConfigFileException(result)) {
    throw new GithubV4Exception(result);
  }

  const { data } = result;

  const sourceCommit = data?.repository?.object;
  if (sourceCommit?.__typename !== 'Commit') {
    throw new BackportError(
      `No commit found on branch "${sourceBranch}" with sha "${sha}"`,
    );
  }

  return parseSourceCommit({ options, sourceCommit });
}
