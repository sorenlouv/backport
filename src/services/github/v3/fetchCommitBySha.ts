import isEmpty from 'lodash.isempty';
import ora from 'ora';
import { BackportOptions } from '../../../options/options';
import { CommitSelected } from '../../../types/Commit';
import { HandledError } from '../../HandledError';
import { getFormattedCommitMessage, getShortSha } from '../commitFormatters';
import { apiRequestV3 } from './apiRequestV3';

export interface GithubCommit {
  commit: {
    message: string;
  };
  sha: string;
}

interface GithubSearch<T> {
  items: T[];
}

export async function fetchCommitBySha(
  options: BackportOptions & { sha: string }
): Promise<CommitSelected> {
  const {
    githubApiBaseUrlV3,
    repoName,
    repoOwner,
    sha,
    accessToken,
    username,
  } = options;

  const spinner = ora(`Loading commit "${getShortSha(sha)}"`).start();

  try {
    const res = await apiRequestV3<GithubSearch<GithubCommit>>({
      method: 'get',
      url: `${githubApiBaseUrlV3}/search/commits?q=hash:${sha}%20repo:${repoOwner}/${repoName}&per_page=1`,
      auth: {
        username: username,
        password: accessToken,
      },
      headers: {
        Accept: 'application/vnd.github.cloak-preview',
      },
    });

    spinner.stop();

    // TODO: it should be possible to backport from other branches than master
    if (isEmpty(res.items)) {
      throw new HandledError(`No commit found on master with sha "${sha}"`);
    }

    const commitRes = res.items[0];
    const fullSha = commitRes.sha;

    const formattedMessage = getFormattedCommitMessage({
      message: commitRes.commit.message,
      sha: fullSha,
    });

    return {
      sourceBranch: 'master',
      selectedTargetBranches: [],
      formattedMessage,
      sha: fullSha,
    };
  } catch (e) {
    spinner.fail();
    throw e;
  }
}
