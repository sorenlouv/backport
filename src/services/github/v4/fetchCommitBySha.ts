import chalk from 'chalk';
import ora from 'ora';
import { BackportOptions } from '../../../options/options';
import { Commit } from '../../../types/Commit';
import { HandledError } from '../../HandledError';
import {
  getFormattedCommitMessage,
  getShortSha,
  getPullNumberFromMessage,
} from '../commitFormatters';
import { apiRequestV4 } from './apiRequestV4';
import {
  PullRequestNode,
  pullRequestFragment,
  pullRequestFragmentName,
  getExistingTargetPullRequests,
  getPullRequestLabels,
} from './getExistingTargetPullRequests';
import { getTargetBranchesFromLabels } from './getTargetBranchesFromLabels';

export async function fetchCommitBySha(
  options: BackportOptions & { sha: string }
): Promise<Commit> {
  const { accessToken, githubApiBaseUrlV4, repoName, repoOwner } = options;

  const query = /* GraphQL */ `
  query CommitsBySha($repoOwner: String!, $repoName: String!, $oid: String!) {
    repository(owner: $repoOwner, name: $repoName) {
      object(expression: $oid) {
        ... on Commit {
          message
          oid
          associatedPullRequests(first: 1) {
            edges {
              node {
                ...${pullRequestFragmentName}
              }
            }
          }
        }
      }
    }
  }

    ${pullRequestFragment}
  `;

  const spinner = ora(`Loading commit "${getShortSha(options.sha)}"`).start();

  let res: DataResponse;
  try {
    res = await apiRequestV4<DataResponse>({
      githubApiBaseUrlV4,
      accessToken,
      query,
      variables: {
        repoOwner,
        repoName,
        oid: options.sha,
      },
    });
    spinner.stop();
  } catch (e) {
    spinner.fail();
    throw e;
  }

  if (!res.repository.object) {
    throw new HandledError(
      `No commit found on branch "${options.sourceBranch}" with sha "${options.sha}"`
    );
  }

  const sha = res.repository.object.oid;
  const commitMessage = res.repository.object.message;
  const pullRequestNode =
    res.repository.object.associatedPullRequests.edges?.[0]?.node;

  const pullNumber =
    pullRequestNode?.number || getPullNumberFromMessage(commitMessage);

  const sourceBranch = pullRequestNode?.baseRefName || options.sourceBranch;

  const formattedMessage = getFormattedCommitMessage({
    message: commitMessage,
    pullNumber,
    sha,
  });

  spinner.stopAndPersist({
    symbol: chalk.green('?'),
    text: `${chalk.bold('Select commit')} ${chalk.cyan(formattedMessage)}`,
  });

  const existingTargetPullRequests = getExistingTargetPullRequests(
    commitMessage,
    pullRequestNode
  );

  const targetBranchesFromLabels = getTargetBranchesFromLabels({
    existingTargetPullRequests,
    branchLabelMapping: options.branchLabelMapping,
    labels: getPullRequestLabels(pullRequestNode),
  });

  return {
    sourceBranch,
    targetBranchesFromLabels,
    sha,
    formattedMessage,
    originalMessage: commitMessage,
    pullNumber,
    existingTargetPullRequests,
  };
}

interface DataResponse {
  repository: {
    object: {
      message: string;
      oid: string;
      associatedPullRequests: {
        edges:
          | {
              node: PullRequestNode;
            }[]
          | null;
      };
    } | null;
  };
}
