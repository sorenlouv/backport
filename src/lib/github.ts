import {
  GithubQuery,
  GithubCommit,
  PullRequest,
  Commit,
  GithubIssue,
  GithubPullRequestPayload,
  GithubIssues,
  GithubApiError
} from '../types/types';

import axios, { AxiosResponse } from 'axios';
import querystring from 'querystring';
import get from 'lodash.get';
import { HandledError } from './errors';

let accessToken: string;
function getCommitMessage(message: string) {
  return message.split('\n')[0].trim();
}

export async function getCommits(
  owner: string,
  repoName: string,
  author: string | null
): Promise<Commit[]> {
  const query: GithubQuery = {
    access_token: accessToken,
    per_page: 20
  };

  if (author) {
    query.author = author;
    query.per_page = 5;
  }

  try {
    const res: AxiosResponse<GithubCommit[]> = await axios(
      `https://api.github.com/repos/${owner}/${repoName}/commits?${querystring.stringify(
        query
      )}`
    );

    const promises = res.data.map(async commit => {
      const sha = commit.sha;
      return {
        message: getCommitMessage(commit.commit.message),
        sha,
        pullRequest: await getPullRequestBySha(owner, repoName, sha)
      };
    });

    return Promise.all(promises);
  } catch (e) {
    throw getError(e);
  }
}

export async function getCommit(
  owner: string,
  repoName: string,
  sha: string
): Promise<Commit> {
  try {
    const res: AxiosResponse<GithubCommit> = await axios(
      `https://api.github.com/repos/${owner}/${repoName}/commits/${sha}?access_token=${accessToken}`
    );
    const fullSha = res.data.sha;
    const pullRequest = await getPullRequestBySha(owner, repoName, fullSha);

    return {
      message: getCommitMessage(res.data.commit.message),
      sha: fullSha,
      pullRequest
    };
  } catch (e) {
    throw getError(e);
  }
}

export async function createPullRequest(
  owner: string,
  repoName: string,
  payload: GithubPullRequestPayload
): Promise<PullRequest> {
  try {
    const res: AxiosResponse<GithubIssue> = await axios.post(
      `https://api.github.com/repos/${owner}/${repoName}/pulls?access_token=${accessToken}`,
      payload
    );
    return {
      html_url: res.data.html_url,
      number: res.data.number
    };
  } catch (e) {
    throw getError(e);
  }
}

export async function addLabels(
  owner: string,
  repoName: string,
  pullNumber: number,
  labels: string[]
) {
  try {
    return await axios.post(
      `https://api.github.com/repos/${owner}/${repoName}/issues/${pullNumber}/labels?access_token=${accessToken}`,
      labels
    );
  } catch (e) {
    throw getError(e);
  }
}

async function getPullRequestBySha(
  owner: string,
  repoName: string,
  commitSha: string
): Promise<number> {
  try {
    const res: AxiosResponse<GithubIssues> = await axios(
      `https://api.github.com/search/issues?q=repo:${owner}/${repoName}+${commitSha}+base:master&access_token=${accessToken}`
    );
    return get(res.data.items[0], 'number');
  } catch (e) {
    throw getError(e);
  }
}

export function setAccessToken(token: string) {
  accessToken = token;
}

function getError(e: GithubApiError) {
  if (e.response && e.response.data) {
    return new HandledError(
      JSON.stringify({ ...e.response.data, axiosUrl: e.config.url }, null, 4)
    );
  }

  return e;
}
