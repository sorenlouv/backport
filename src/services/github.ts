import {
  GithubApiError,
  GithubCommit,
  GithubIssue,
  GithubQuery,
  GithubSearch
} from '../types/GithubApi';
import axios, { AxiosResponse } from 'axios';
import querystring from 'querystring';
import get from 'lodash.get';
import isEmpty from 'lodash.isempty';
import { HandledError } from './HandledError';
import { getPullRequestPayload } from '../steps/doBackportVersions';

export interface Commit {
  sha: string;
  message: string;
  pullNumber?: number;
}

let accessToken: string;
function getCommitMessage(message: string) {
  return message.split('\n')[0].trim();
}

export async function fetchCommitsByAuthor(
  owner: string,
  repoName: string,
  author: string | null
): Promise<Commit[]> {
  const query: GithubQuery = {
    access_token: accessToken,
    per_page: 10
  };

  if (author) {
    query.author = author;
    query.per_page = 5;
  }

  try {
    const res: AxiosResponse<GithubCommit[]> = await axios(
      `https://api.github.com/repos/${owner}/${repoName}/commits?${querystring.stringify(
        query
      )}`
    );

    const promises = res.data.map(async commit => {
      const sha = commit.sha;
      return {
        message: getCommitMessage(commit.commit.message),
        sha,
        pullNumber: await fetchPullRequestNumberBySha(owner, repoName, sha)
      };
    });

    return Promise.all(promises);
  } catch (e) {
    throw getError(e);
  }
}

export async function fetchCommitBySha(
  owner: string,
  repoName: string,
  sha: string
): Promise<Commit> {
  try {
    const res: AxiosResponse<GithubSearch<GithubCommit>> = await axios(
      `https://api.github.com/search/commits?q=hash:${sha}%20repo:${owner}/${repoName}&per_page=1&access_token=${accessToken}`,
      {
        headers: {
          Accept: 'application/vnd.github.cloak-preview'
        }
      }
    );

    if (isEmpty(res.data.items)) {
      throw new HandledError(`No commit found for SHA: ${sha}`);
    }

    const commitRes = res.data.items[0];
    const fullSha = commitRes.sha;
    const pullNumber = await fetchPullRequestNumberBySha(
      owner,
      repoName,
      fullSha
    );

    return {
      message: getCommitMessage(commitRes.commit.message),
      sha: fullSha,
      pullNumber
    };
  } catch (e) {
    throw getError(e);
  }
}

async function fetchPullRequestNumberBySha(
  owner: string,
  repoName: string,
  commitSha: string
): Promise<number> {
  try {
    const res: AxiosResponse<GithubSearch<GithubIssue>> = await axios(
      `https://api.github.com/search/issues?q=repo:${owner}/${repoName}+${commitSha}+base:master&access_token=${accessToken}`
    );
    return get(res.data.items[0], 'number');
  } catch (e) {
    throw getError(e);
  }
}

export async function createPullRequest(
  owner: string,
  repoName: string,
  payload: ReturnType<typeof getPullRequestPayload>
) {
  try {
    const res: AxiosResponse<GithubIssue> = await axios.post(
      `https://api.github.com/repos/${owner}/${repoName}/pulls?access_token=${accessToken}`,
      payload
    );
    return {
      html_url: res.data.html_url,
      number: res.data.number
    };
  } catch (e) {
    throw getError(e);
  }
}

export async function addLabelsToPullRequest(
  owner: string,
  repoName: string,
  pullNumber: number,
  labels: string[]
) {
  try {
    return await axios.post(
      `https://api.github.com/repos/${owner}/${repoName}/issues/${pullNumber}/labels?access_token=${accessToken}`,
      labels
    );
  } catch (e) {
    throw getError(e);
  }
}

export function setAccessToken(token: string) {
  accessToken = token;
}

function getError(e: GithubApiError) {
  if (e.response && e.response.data) {
    return new HandledError(
      JSON.stringify({ ...e.response.data, axiosUrl: e.config.url }, null, 4)
    );
  }

  return e;
}
