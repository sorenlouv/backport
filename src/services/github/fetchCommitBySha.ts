import axios from 'axios';
import isEmpty from 'lodash.isempty';
import { BackportOptions } from '../../options/options';
import { HandledError } from '../HandledError';
import { CommitSelected } from './Commit';
import { GithubCommit, GithubSearch } from './GithubApiTypes';
import { handleGithubError } from './handleGithubError';
import { getFormattedCommitMessage } from './commitFormatters';

export async function fetchCommitBySha(
  options: BackportOptions & { sha: string }
): Promise<CommitSelected> {
  const { apiHostname, repoName, repoOwner, sha, accessToken } = options;
  try {
    const res = await axios.get<GithubSearch<GithubCommit>>(
      `https://${apiHostname}/search/commits?q=hash:${sha}%20repo:${repoOwner}/${repoName}&per_page=1&access_token=${accessToken}`,
      {
        headers: {
          Accept: 'application/vnd.github.cloak-preview'
        }
      }
    );

    if (isEmpty(res.data.items)) {
      throw new HandledError(`No commit found on master with sha "${sha}"`);
    }

    const commitRes = res.data.items[0];
    const fullSha = commitRes.sha;

    const message = getFormattedCommitMessage({
      message: commitRes.commit.message,
      sha: fullSha
    });

    return {
      message,
      sha: fullSha
    };
  } catch (e) {
    throw handleGithubError(e);
  }
}
