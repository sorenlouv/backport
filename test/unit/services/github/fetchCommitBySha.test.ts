import axios from 'axios';
import { fetchCommitBySha } from '../../../../src/services/github/fetchCommitBySha';
import { getDefaultOptions } from './getDefaultOptions';

describe('fetchCommitBySha', () => {
  it('should return single commit with pull request', async () => {
    const commitSha = 'myCommitSha';
    const options = getDefaultOptions();

    const axiosSpy = jest
      .spyOn(axios, 'get')

      // mock commits
      .mockResolvedValueOnce({
        data: { items: [{ commit: { message: 'myMessage' }, sha: commitSha }] }
      })

      // mock PRs
      .mockResolvedValueOnce({
        data: { items: [{ number: 'myPullRequestNumber' }] }
      });

    expect(await fetchCommitBySha({ ...options, sha: commitSha })).toEqual({
      message: 'myMessage (#myPullRequestNumber)',
      pullNumber: 'myPullRequestNumber',
      sha: 'myCommitSha'
    });

    expect(axiosSpy.mock.calls).toMatchSnapshot();
  });
});
