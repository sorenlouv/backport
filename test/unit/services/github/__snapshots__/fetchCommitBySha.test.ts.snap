// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchCommitBySha should return single commit with pull request 1`] = `
Array [
  Array [
    "https://api.github.com/search/commits?q=hash:myCommitSha%20repo:elastic/kibana&per_page=1&access_token=myAccessToken",
    Object {
      "headers": Object {
        "Accept": "application/vnd.github.cloak-preview",
      },
    },
  ],
  Array [
    "https://api.github.com/search/issues?q=repo:elastic/kibana+myCommitSha+base:master&access_token=myAccessToken",
  ],
]
`;
